### ISSUE TITLE
.NET 6?

### ISSUE THREAD

**Issue by jensbrak** [Wufflez/Loki#17]
I am playing with some modifications wto Loki (neat project btw - I'm more interested in it from a code perspective rather than actually use it for its intended purpose). Maybe I can do some PR's with what I do or might do, if that's accepted or even appreciated.

First up a question though: any reason for sticking to .NET Core 3.1? 
(I can think of many but for me it was just natural to go to .NET 6 when I first found and tried out this project. Just so that I know if I should avoid dependencies to anything > .NET Core 3.1)

**Comment by jensbrak** [comment]
Well, my first PR would be .NET 6 and adding some missing items. I also
refactored the Shared Item Database so that it is read from a CSV file
rather than being hard coded. I chose CSV rather than Json or any other
format since CSV allows for easy imports/exports from/to Excel, Google
Sheets or any other source in the future. These changes are already done so
a PR is just a matter of submitting.

The editor works with latest version of the game. However, part from
missing items something seems off with item repair. It does not affect the
saved file butt does not work (for me at least) on items that needs repair.

An annoying feature would be that the default selection of profile is the
backup copy rather than the original profile file. Not a bug but a...
feature :)

As for features I can think of several. For instance... Drag and drop
order  of items. Possibility to select equipped items, not only visualize
them, item macro buttons (something to save a different custom made item
collections to quickly replace the current load out with).

I found this project since I was curious of the profile save file format. I
kinda like to fiddle with things. How did you deduce the format? Is there
any good source of information (part from your code that is) for the
format? The item properties can be found but I have not found a source that
is both up to date AND compact, yet.

Den fre 18 feb. 2022 19:57Adam ***@***.***> skrev:

**Comment by Wufflez** [comment]
>The editor works with latest version of the game. However, part from
>missing items something seems off with item repair. It does not mess up the
>saved file but does not work  either (for me at least) on items that needs repair.

Yeah thanks for pointing that out. I saw your pull request to fix it but it was kinda a band aid over the real problem which stems from a previous update to the editor from a few releases ago. I've fixed that now and will make a release very soon with that in that should resolve it.

>An annoying feature (?...) would be that the default selection of profile is the
>backup copy rather than the original profile file. Not a bug but a...
>feature , right? :)

I'm not sure what you mean by this, as the backups are not directly selectable by default beacuse they go into a subfolder.
The default *should* be just the first profile that exists alphabetically which is a little weird and should maybe be the most recently modified file (i.e. most recently played one). I'll look at changing that!

>I found this project since I was curious of the profile save file format. I
>kinda like to fiddle with things. How did you deduce the format? Is there
>any good source of information (part from your code that is) for the
>format? The item properties can be found but I have not found a source that
>is both up to date AND compact, yet.

I deduced the format by inspecting the source code for the game. There are countless tools out there to reverse engineer .NET assemblies but if you're unfamiliar and just want to look at what I used - then take a look at [dnSpy ](https://github.com/dnSpy/dnSpy) as it's probably still the fastest and easiest tool to not only reverse engineer but debug production .NET binaries with. It is archived, sadly, without really any obvious reason as to why!

There are a number of binary 'blobs' for lack of a better word that make up the player file. The blobs that are subject to change have a version number and you'll notice that the game client does its best to carefully check older files and load default values for new properties or data and then always saves the files as the most recent format version. 

I have done my best to replicate the same logic as the game client, and simply play it safe if a version is not recognised and hardcoded in the supported version list then Loki will refuse to try to convert it. This should really only happen if somebody somehow had a version from a super early internal pre-release or has a file from a very new release of the game that I haven't seen before. Any data blobs that aren't supported (yet) by Loki for example the world data - including the 'fog of war' map reveal etc are just kept in memory and written out byte-for-byte as they were so don't really cause any issue. Care does need to be taken to make sure anything new that creeps in like this is accounted for. There is a total data length value that acts as another safeguard for potential data blocks being added and then ignored by Loki, which would cause an error when loading a profile as we carefully check how many bytes we have processed vs how many the file says we should have read.

I'm not aware of any other source for the format, but to be honest I haven't googled it since writing the tool. There wasn't anything obvious around at the time but there may well be some upto date and easier to read documentation on the save format somewhere. The game client code is fairly easy to read if you don't mind the strange variable names that the decompiler chooses to use.

**Comment by jensbrak** [comment]
> Yeah thanks for pointing that out. I saw your pull request to fix it but it was kinda a band aid over the real problem which stems from a previous update to the editor from a few releases ago. I've fixed that now and will make a release very soon with that in that should resolve it.

I knew it was a quick fix but worked good enough for me to stop digging. (Or not. I digged allright but stopped only to realize my confidence in xaml-land is not in par with my curiousity and need to fiddle around. I'm an old Forms dude when it comes to .net desktop.)

> I'm not sure what you mean by this

Can't blame you, it was me mixing things up. I believe the "extra profile" was a left over from another program I tried that made backup copies that where synced back with Google Drive or someting like that. Bottom line: it had nothing to do with this program - which I suspected but couldn't wrap my head around.

> I deduced the format by inspecting the source code for the game

Interesting! Of course, when you say it it makes sense.  I knew it's a .NET application but never connected the dots. I was so busy finding reference documentation of some sorts, unofficial or not. :)

Thanks for taking time to kindly adress my... twaddle. :) Updating to .NET 6 was a a breeze.

**Comment by jensbrak** [comment]
Oh. I did write some more tests and fixes - including "repair all" button and some other things. However, I did have a feeling that things would be complicated adding those - since much of the code seemed like in the middle of some refactoring. Or if it was me remembering why I prefer Forms over XAML (even though I know the latter has some really neat features!). :)

Anyways, thanks for a fund and interesting piece of software to fiddle around with! I almost ended up diving into the XAML world (finally - been close before) but escaped narrowly ;)
